### ISSUE TITLE
Have webapi send JWT to Sipi for auth

### ISSUE THREAD

**Issue by benjamingeer** [dhlab-basel/Knora#520]
The Knora API server can use JWT to authorise a user to upload files to Sipi, as well as to view files stored in Sipi.

**Comment by benjamingeer** [comment]
It seems that JWT isn't going to work, because of HTML `img` tags, so it looks like Sipi will need to make a request to the Knora API server for this.

**Comment by subotic** [comment]
relevant: #171

**Comment by subotic** [comment]
As I understand it, this is only a problem in Salsah 1.5, where the `img` tag is used. In Salsah 2.0 this shouldn't be a problem. I will talk to @lrosenth about this.

**Comment by lrosenth** [comment]
I think SIPI is accepting webtokens as cookie or as URL option (don't remember the "keyword" to be used...). May be this can be used for Salsah 1.5 bay adding the token to the URL in the img-ag.

Sent from my iPad

On 9 Aug 2017, at 11:00, Ivan Subotic <notifications@github.com<mailto:notifications@github.com>> wrote:


As I understand it, this is only a problem in Salsah 1.5, where the img tag is used. In Salsah 2.0 this shouldn't be a problem. I will talk to @lrosenth<https://github.com/lrosenth> about this.

**Comment by subotic** [comment]
> May be this can be used for Salsah 1.5 bay adding the token to the URL in the `img-tag`.

Yes, this is what I also had in mind. We could generate so called `access tokens` and add them to every image URL. Those access tokens would only be valid for one particular image and only for a short amount of time. The pre-flight script in SIPI could then check the validity of the token and give access, all without the need to contact `webapi`.

**Comment by benjamingeer** [comment]
Just talked to Lukas about this again. Here's the problem:

Putting a JWT in an image URL is fine for SALSAH 1.5, but it won't work for a IIIF viewer. The IIIF auth standard expects to get a JWT in a cookie or in an HTTP auth header. There's no way to set an auth header for an `<img>` tag, so the cookie is the only option. But we can't make one giant cookie containing the user's permissions on all images. We can make a cookie that says who the user is, but then Sipi still has to make a request to Knora to ask for the user's permissions on a particular image. And then we're back where we started.

So my question is this: If the purpose of using JWT is to avoid the Sipi-to-Knora request, but we're actually going to have to make that request anyway, why go to the trouble of using JWT? Why not just keep using cookies as we do now?

**Comment by subotic** [comment]
As discussed with @benjamingeer, I will take a look at the IIIF standard and [OpenSeadragon](https://openseadragon.github.io)  to see if and how we can use JWT, and then decide from then on.

**Comment by subotic** [comment]
There seems to be a nice solution with [Angular HTTP Interceptors](https://angular.io/guide/http#intercepting-all-requests-or-responses) (provided in Angular 4.3). As I understand it, it should be possible to intercept each call from SALSAH to SIPI and remove the token from the URL and add it to the header (see: https://github.com/dougmoscrop/angular-img-http-src).

If the HTTP interceptor idea doesn't work, then we will proceed as recently discussed with @lrosenth, and "break" the Authentication API 1.0 by using tokens attached to the URL. This will also mean, that we will maybe need to patch OpenSeadragon, in the case that it doesn't leave the tokens.

**Comment by kilchenmann** [comment]
In Salsah resp. in the knora-ui modules we implemented the Angular HTTP Interceptors and I hope we can use it for the Sipi authentication as well

**Comment by benjamingeer** [comment]
@mrivoal In API v2, the user submits a JWT token to Sipi when uploading a file:

https://docs.knora.org/paradox/03-apis/api-v2/editing-values.html#creating-file-values

We haven't yet implemented using JWT to authorise a user to view a file. Is it correct that you need this feature for your release at the end of February?

**Comment by gfoo** [comment]
> We haven't yet implemented using JWT to authorise a user to view a file. Is it correct that you need this feature for your release at the end of February?

yes, I get an image this way: `http://localhost:1024/knora/AU8mw3gyud-GKKZruS24Z9.jpx/full/full/0/default.jpg` and thanks to Angular Interceptor, I add an auth method in HTTP Header: `Authorization: Bearer <token>`.

n.b.: I have a solution to combine easily `<img>` and `Authorization: Bearer <token>` , let me know if you are interested

**Comment by benjamingeer** [comment]
> n.b.: I have a solution to combine easily <img> and Authorization: Bearer <token> , let me know if you are interested

Yes please!

**Comment by gfoo** [comment]
https://stackoverflow.com/questions/46563607/angular-4-image-async-with-bearer-headers

By using an Angular Pipe allowing an asynchronous loading of the image content with the http client which depends on the interceptor (which adds the bearer auth conf):

```typescript
@Pipe({
  name: 'AsyncLoadingImage'
})
export class AsyncLoadingImagePipe implements PipeTransform {
  constructor(private http: HttpClient) {}

  transform(url: string) {
    return this.http.get(url, { responseType: 'blob' }).pipe(
      switchMap(blob => {
        // return new observable which emits a base64 string when blob is converted to base64
        return Observable.create(observer => {
          const reader = new FileReader();
          reader.readAsDataURL(blob); // convert blob to base64
          reader.onloadend = function() {
            observer.next(reader.result); // emit the base64 string result
          };
        });
      })
    );
  }
}
```
In the html:
```html
<img *ngIf="sipiImageURL | AsyncLoadingImage | async as imgSrc" [attr.src]="imgSrc">
```

Can probably done from pure javascript too.

**Comment by subotic** [comment]
cool, thanks. I think @kilchenmann will like this a lot. We still need to add this to Salsah.

**Comment by benjamingeer** [comment]
@gfoo If you do it this way, I guess Knora has to provide the token separately from the image URL. Maybe something like this?

```jsonld
  "knora-api:hasStillImageFileValue" : [ {
    "@id" : "http://rdfh.ch/7bbb8e59b703/reps/22bc3f713f07",
    "@type" : "knora-api:StillImageFileValue",
    "knora-api:attachedToUser" : {
      "@id" : "http://rdfh.ch/users/91e19f1e01"
    },
    "knora-api:fileValueAsUrl" : {
      "@type" : "xsd:anyURI",
      "@value" : "http://localhost:1024/knora/incunabula_0000004159.jp2/full/3625,5251/0/default.jpg"
    },
    "knora-api:fileValueHasToken" : "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJLbm9yYSIsInN1YiI6Imh0dHA6Ly9yZGZoLmNoL3VzZXJzLzlYQkNyRFYzU1JhN2tTMVd3eW5CNFEiLCJhdWQiOlsiS25vcmEiLCJTaXBpIl0sImV4cCI6NDY5NTE5MzYwNSwiaWF0IjoxNTQxNTkzNjA1LCJqdGkiOiJsZmdreWJqRlM5Q1NiV19NeVA0SGV3IiwiZm9vIjoiYmFyIn0.qPMJjv8tVOM7KKDxR4Dmdz_kB0FzTOtJBYHSp62Dilk",
    "knora-api:fileValueHasFilename" : "incunabula_0000004159.jp2",
    "knora-api:hasPermissions" : "CR knora-base:Creator|M knora-base:ProjectMember|V knora-base:KnownUser|RV knora-base:UnknownUser",
    "knora-api:stillImageFileValueHasDimX" : 3625,
    "knora-api:stillImageFileValueHasDimY" : 5251,
    "knora-api:stillImageFileValueHasIIIFBaseUrl" : {
      "@type" : "xsd:anyURI",
      "@value" : "http://localhost:1024/knora"
    }
  }
```

**Comment by gfoo** [comment]
Not sure to understand, I get the token when I connect to Knora with `/v2/authentication`, why do you want to provide it when we get the `knora-api:hasStillImageFileValue`?

**Comment by benjamingeer** [comment]
When you log in, you get a token that just says you're logged in. But it can't say anything about image permissions, because every image can have different permissions. Therefore, every image needs a different token.

**Comment by benjamingeer** [comment]
The token for an image needs to specify which permissions you have on that particular image.

**Comment by gfoo** [comment]
So, in the previous solution there was a cookie per image?

**Comment by gfoo** [comment]
I thought that Sipi have to ask to Knora if the user (represented by the token) is allowed to see or not an image (represented by its sipi id), in this case, no need for a token per image?

**Comment by loicjaouen** [comment]
that was the case in v1, a session id was added in cookies, then sipi was forwarding it to knora to ask for permissions.

The JWT is supposed to give the access right information to sipi without sipi having to ask knora.

**Comment by benjamingeer** [comment]
in the previous solution, every time you request an image from Sipi, the browser sends your login cookie to Sipi, and Sipi makes an HTTP request to Knora (using the session ID from your login cookie), to ask what permissions you have on that particular image. But this is inefficient, because it requires an extra HTTP request every time you request an image. This is why we want to use JWT instead: Knora will serve the token with the image, the client will give the token to Sipi, and the token will tell Sipi what the user's permissions are on the image. This eliminates the need for an extra HTTP request.

**Comment by gfoo** [comment]
But now the Angular interceptor is no more interesting then... the idea that was to take care behind the scene of a global token precisly!

`"knora-api:fileValueHasToken"`: ok fine

**Comment by benjamingeer** [comment]
> But now the Angular interceptor is no more interesting then... the idea that was to take care behind the scene of a global token precisly!

I thought the idea was to be able to add an authentication header (containing the token) to the HTTP request that is sent because of the `<img>` tag. This is the reason why it has taken so long to resolve this issue. To make the token work with `<img>` tags, it seemed that the only option was to put the token in the image URL, like this:

```
http://localhost:1024/knora/incunabula_0000004159.jp2/full/3625,5251/0/default.jpg?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJLbm9yYSIsInN1YiI6Imh0dHA6Ly9yZGZoLmNoL3VzZXJzLzlYQkNyRFYzU1JhN2tTMVd3eW5CNFEiLCJhdWQiOlsiS25vcmEiLCJTaXBpIl0sImV4cCI6NDY5NTE5MzYwNSwiaWF0IjoxNTQxNTkzNjA1LCJqdGkiOiJsZmdreWJqRlM5Q1NiV19NeVA0SGV3IiwiZm9vIjoiYmFyIn0.qPMJjv8tVOM7KKDxR4Dmdz_kB0FzTOtJBYHSp62Dilk
```

But this is non-standard in IIIF, so the question is whether this will work with IIIF viewers.

I thought that your Angular interceptor would solve this problem by putting the token in a header, so it doesn't need to be in the URL.

**Comment by gfoo** [comment]
Interceptor is not related to authentication header. It works behind the scene for every http request without knowing the context, this is nice for the Knora API to automatically add a token in the header.

For image from Sipi I won't use the interceptor, for each http request I will add the particular token each time. Just few modifications in my angular pipe, no problem.

**Comment by benjamingeer** [comment]
> For image from Sipi I won't use the interceptor, for each http request I will add the particular token each time. Just few modifications in my angular pipe, no problem.

Could you give an example of how to do this with an IIIF viewer? I would like to make sure whichever approach we decide to use will work with IIIF viewers like https://openseadragon.github.io/ and http://projectmirador.org/. I am assuming that the IIIF viewer is not going to know about Angular, so I wonder if Angular pipes can actually solve this problem.

**Comment by musicEnfanthen** [comment]
Just to make sure, I understand it right: Finally the Client has to check the permissions for every file himself instead of Knora overtaking this task?

**Comment by benjamingeer** [comment]
The token will basically say something like: “I, Knora, have determined that the user carrying this token has permission to view file xyz.jp2.”

**Comment by benjamingeer** [comment]
The only thing that changes is how Knora communicates this information to Sipi.

**Comment by gfoo** [comment]
> Could you give an example of how to do this with an IIIF viewer? I would like to make sure whichever approach we decide to use will work with IIIF viewers like https://openseadragon.github.io/ and http://projectmirador.org/. I am assuming that the IIIF viewer is not going to know about Angular, so I wonder if Angular pipes can actually solve this problem.

My solution is strictly Angular oriented and you must have a deep access of the code if you want to hack an existing software.

I never used openseadragon or projectmirador, if the common piece is the IIIF API, the logical solution is to implement the IIIF Auth API (see @subotic above comment), but don't know if it covers your requirements.

A short time and straightforward solution is clearly to add `?token=<token>` because all informations are in the URL, not sure that it really breaks the IIIF API, it is an optional parameter after all :)

**Comment by benjamingeer** [comment]
> if the common piece is the IIIF API, the logical solution is to implement the IIIF Auth API (see @subotic above comment), but don't know if it covers your requirements.

According to Lukas, it doesn't have the concept of different permissions for each image. We have been talking about suggesting an enhancement to the IIIF consortium to allow for this. But that will take time.

> A short time and straightforward solution is clearly to add `?token=<token>` because all informations are in the URL, not sure that it really breaks the IIIF API, it is an optional parameter after all :)

We have to try it and see if the IIF viewers actually accept it and if they pass it to the IIIF server (Sipi).

**Comment by musicEnfanthen** [comment]
Does https://iiif.io/api/auth/1.0/#example-description-resource-with-authentication-services and point 3.0 below help?

**Comment by benjamingeer** [comment]
I don't think it helps, because as far as I can tell, the IIIF Authentication API is based on the idea that the user logs in and gets a single token that they will use for many different images. The documentation doesn't seem to describe a use case like ours, where the user gets a different token for each image requested.

**Comment by musicEnfanthen** [comment]
Thanks @benjamingeer for your explanations. I confess, I am still a little bit confused about the different parts/applications involved here and their role in the operational flow. Possibly it is my bad understanding, I apologize.

Why does the IIIF viewer have to know anything about authentication when that is handled by the client, so Angular in these cases? The Angular app checks the auth and, if allowed, sends the image to the view, so the IIIF viewer. Probably I miss something here.

@gfoo Thanks for the approach with interceptors. But don't know if making an http request from within a pipe is a good idea? Couldn't it be done as well from within a service that handles all http stuff and just returns the correct `imageSrc` to the view to be displayed in `img` tag?

**Comment by benjamingeer** [comment]
> Why does the IIIF viewer have to know anything about authentication when that is handled by the client, so Angular in these cases? The Angular app checks the auth and, if allowed, sends the image to the view, so the IIIF viewer. Probably I miss something here.

The Angular app doesn't get the image, the IIIF viewer does. The Angular app is just responsible for providing the IIIF viewer with a JSON document describing the image. The IIIF viewer is then responsible for getting the image (or parts of it) directly from an IIIIF server. To do this, the IIIF viewer has to generate HTML (presumably using `<img>` tags) that causes the browser to make an HTTP request to the IIIF server (Sipi). So if we are going to use a token as a credential, the IIIF viewer somehow has to have this token and include it in the request to Sipi.

Try viewing the source code of this page to see an example:

https://showcase.iiif.io/showcase/osd-viewer/

**Comment by gfoo** [comment]
Note that we of course don't need anymore of `AsyncLoadingImagePipe` with the `?token` solution:
```html
<img src="sipiImageURL+'?token='+sipiImageToken" >
```
And, also get a direct access even from your browser to the image just by using the built URL, can be more efficient, for debug, for internal browser cache image...

@musicEnfanthen with a `Bearer` auth way, the token is not in the imageSrc value but in the HTTP header request, can not be done directly from your browser for example.

**Comment by benjamingeer** [comment]
@musicEnfanthen Also, keep in mind that Knora and Sipi cannot trust client code to check authorisation (anyone could write a client that simply allows everything), so the checking has to be done on the server.

**Comment by subotic** [comment]
I could but not before next week. Would like to first finish the non-actor responders.

Although, I don't really know how. I should probably first try openseadragon and mirador with `?token=xyz`in the image URL. If those work, then we could implement it this way.

**Comment by benjamingeer** [comment]
I asked about this on the IIIF-Discuss mailing list, and got some interesting replies. See especially the reply from Tom Crane, Technical Director at Digerati, the company that makes the IIIF viewer [Universal Viewer](https://digirati.com/work/galleries-libraries-archives-museums/products/universal-viewer/):

https://groups.google.com/forum/#!topic/iiif-discuss/f9-eQbdqgME

**Comment by lrosenth** [comment]
Very, very interesting!

Thanks for sharing. We’ll hve to discuss this!

Best Regards,
Lukas (still at home...)

Sent from my iPad

On 11 Jan 2019, at 11:01, Benjamin Geer <notifications@github.com<mailto:notifications@github.com>> wrote:


I asked about this on the IIIF-Discuss mailing list, and got some interesting replies. See especially the reply from Tom Crane, Technical Director at Digerati, the company that makes the IIIF viewer Universal Viewer<https://digirati.com/work/galleries-libraries-archives-museums/products/universal-viewer/>:

https://groups.google.com/forum/#!topic/iiif-discuss/f9-eQbdqgME

**Comment by benjamingeer** [comment]
@subotic I think Knora-ui actually generates `info.json` in Angular, based on the information from the Knora API, but you'd have to ask the Knora-ui people about that.

The key question seems to be whether IIIF viewers can actually handle a token in the image URL.

**Comment by benjamingeer** [comment]
Now Andy Irving at the British Library has replied to the mailing list thread, saying "Feel free to get in touch off list if you want any more information". So I think we should just ask him for details about how they're doing this.

**Comment by subotic** [comment]
Andy Irving wrote:
> technically we support different rights in a region of an image but thankfully we don't currently have any of those in use.

This is on a whole other level.

Then I have understood it correctly. They are doing what I have described in my previous comment.

**Comment by benjamingeer** [comment]
In his latest message he says:

> Sorry for not being clear! When the client makes the XHR call to the info.json, it has to use the token because it can't access the correct cookie. when making the requests for pixels, it can send the cookies assigned during the login action. 
>
> We use SAML2 (shibboleth) as our authentication system, and it sets a client cookie, which gets sent with the request for pixels. our image server is proxied by a shibbolth service provider that uses that cookie to set upstream data (role). our rights enforcement is done at that point.

So they're still using a cookie in the request to get the image file.

**Comment by subotic** [comment]
So they intercept each call to the IIIF server, set the authorization, and then the IIIF server can act accordingly:

client -> shibboleth (proxy) -> image server

**Comment by subotic** [comment]
we could implement it by not allowing direct access to Sipi, but only through Knora which will then talk to Sipi. Basically, create an authorization proxy route.

**Comment by benjamingeer** [comment]
> we could implement it by not allowing direct access to Sipi, but only through Knora which will then talk to Sipi. Basically, create an authorization proxy route.

But that would still require an extra HTTP request (and triplestore query) per image, which is what we have already. The only difference would be the order of the requests: client -> Knora -> Sipi instead of the current client -> Sipi -> Knora.

**Comment by subotic** [comment]
After the token expires (10 minutes or so), the tokens would need to be refreshed somehow. We need a per image authorization route anyway. The question is who is going to do the refreshing.

1. If we use the IIIF Authorization API then the client (IIIF viewer) needs to implement this feature, which if they have implemented the Authorization API they should have already done so.

2. Let's assume viewers work with appended JWT tokens to the image URL, then the viewer will work for 10 minutes (or whatever the lifetime of the token is). After that, the whole page needs to be refreshed so that all image URLs with fresh tokens can be requested.

In case 1 only those tokens will be refreshed (and also initially requested), for images that are actually viewed.
In case 2 all image tokens need to be refreshed every single time.

I think that case 1 is going to generating less traffic.

**Comment by subotic** [comment]
in case (1) the viewer would go to the authentication endpoint specified in the manifest file to get the token and then use this token to access Sipi. Since we apparently can set a different authentication endpoint for each image, we can generate a token with permissions specific to each image.

Now the most important thing is that the viewer handles token expiration, i.e. re-authenticates itself for every image when the token expires.

**Comment by subotic** [comment]
maybe a quick Skype call?

**Comment by benjamingeer** [comment]
Yes, I’ll call you in 20 minutes or so.

**Comment by benjamingeer** [comment]
Sorry, that didn't work out, let's talk on Monday.

**Comment by subotic** [comment]
So it looks like that using the IIIF Authentication API is not going to be trivial. After talking to @benjamingeer, we decided that the fastest way would be to use session cookies like in `Salsah1.5`, and let Sipi ask Knora for the user's permissions for each requested image. Knora will cache this information so that the load on the triplestore should not be too high.

The implementation would involve changing `V2` authentication to also create a session cookie. Then all HTML `img` tag requests should automatically include this information when talking to Sipi.

@gfoo would this be alright with you for your February release?

**Comment by subotic** [comment]
I‘ve read the IIIF Authentication API spec and @benjamingeer you where right. Both cookies and tokens are used. **Cookies** for **Content Resources** (image) and **tokens** for **Description Resources** (info,json).

So our proposed first step of adding session creation to `V2` authentication alongside token is completely in line with the IIIF spec.

**Comment by gfoo** [comment]
So it means that I can use a Sipi image URL directly in the html `img src` attribute if I use the `v2 auth` endpoint (what I do actually) ?

**Comment by gfoo** [comment]
Sipi and Knora being not the same server you are taking about [Third-party cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#Third-party_cookies)?

**Comment by subotic** [comment]
> So it means that I can use a Sipi image URL directly in the html `img src` attribute if I use the `v2 auth` endpoint (what I do actually) ?

Yes, that would be the goal.

> Sipi and Knora being not the same server you are taking about [Third-party cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#Third-party_cookies)?

The Sipi and Knora API endpoint need to be under the same domain, e.g., `api.example.com` and `iiif.example.com`. This is how we already deploy.

**Comment by benjamingeer** [comment]
> The Sipi and Knora API endpoint need to be under the same domain

I thought the client had to log into Knora first, then log into Sipi, so Sipi could set a cookie containing the same session ID. Isn't this what the Sipi script `Knora_login.lua` is for?

**Comment by subotic** [comment]
>  thought the client had to log into Knora first, then log into Sipi, so Sipi could set a cookie containing the same session ID. Isn't this what the Sipi script Knora_login.lua is for?

Not anymore. I changed it in November or so for `V1`. The client only needs to login into Knora and the same cookie will then be used by Sipi.

The `Knora_login.lua` and `Knora_logout.lua` are not used anymore. The calls to those routes are commented out in `Salsah1`.

**Comment by benjamingeer** [comment]
OK great. In that case, could you please delete those scripts?

**Comment by subotic** [comment]
If somebody, for whatever reason, want's to have API on one domain and SIpi on another, then he would/could theoretically use those. But I don't think that we need to support this use case.

**Comment by subotic** [comment]
I will delete them.

**Comment by subotic** [comment]
Yes, exactly.

**Comment by gfoo** [comment]
Just tested #1165 and it works fine (don't know if Sipi correctly use the cookies sent by the browser to check the permissions but I can now see correctly the Sipi images from my Angular app)

**Comment by subotic** [comment]
Sipi should already use it. It is the way the `Salsah1.5` GUI is working.

**Comment by gfoo** [comment]
Ok, it works from Salsah (dev stack) after having manually removed the cookie because logout seems to not do that, is something that I have to do (remove cookie) from my app when I logout from Knora using `DELETE /v2/authentication` or is it the Knora's responsibility?

**Comment by subotic** [comment]
> or is it the Knora's responsibility?

Yes, Knora should send the same cookie with the changed expire date. What I have found, is that with Safari I need to refresh the browser to see that the cookie was removed.

**Comment by subotic** [comment]
> where I expected and additional ; Domain=.example.com

Good catch. This needs to be set. The current implementation wouldn't work in production.

**Comment by subotic** [comment]
> > where I expected and additional ; Domain=.example.com
> 
> Good catch. This needs to be set. The current implementation wouldn't work in production.

should be fixed in #1170

**Comment by gfoo** [comment]
finally, not so sure that it works very well with Angular through their HttpClient, it worked a first time but not anymore :(

**Comment by subotic** [comment]
Cookies are a browser functionality. I don't think that it matters if Angular is used or not. Which browser are you using for testing? I find Safari better for the cookie stuff, as you can see the whole HTTP response header including the set-cookie header.

**Comment by gfoo** [comment]
If you want to quickly test login/logout with Angular without dev env: 
 https://stackblitz.com/edit/angular-ea4gsz

**Comment by subotic** [comment]
This seems to work. I had to download and run it locally, because of the blocked mixed content (stackblitz is https and localhost is http).

You could make another button to check with Knora: `GET /v2/authentication` tries to authenticate the user and tells you the result.

**Comment by gfoo** [comment]
I updated the project, I can see the `Set-Cookie: KnoraAuthentication=eyJ...` in the response header by my browser definitely does not store the cookie.
<img width="1274" alt="capture d ecran 2019-01-17 a 18 15 07" src="https://user-images.githubusercontent.com/15976605/51336023-dc441c80-1a83-11e9-9634-4874f9bfc5ab.png">
<img width="1262" alt="capture d ecran 2019-01-17 a 18 13 47" src="https://user-images.githubusercontent.com/15976605/51336037-e403c100-1a83-11e9-8278-22051e4eae0b.png">

**Comment by subotic** [comment]
I think that the cookie is there. Just to see it, you need to refresh the page.

**Comment by gfoo** [comment]
This is not related directly to Angular HttpClient but to XMLHttpRequest, see this test: https://gist.github.com/gfoo/6d9f449b55d4ae19fe8ecfcbfa1f70d8
For me, with or without `withCredentials=true` makes a difference.

**Comment by subotic** [comment]
> withCredentials=true

Ah, yes. Now that you mention it. I have fallen for this one in the past. We have this set in `Salsah1` (https://github.com/dhlab-basel/Knora/blob/ac2c9bafc4d211f126660cbf10e9f8330b1cb614/salsah1/src/public/js/01_salsah_api.js#L60-L62)

**Comment by benjamingeer** [comment]
Would it help to document that somewhere?

**Comment by subotic** [comment]
@kilchenmann For a future reference, don't forget to set `withCredentials=true`, when dealing with cookies.

**Comment by subotic** [comment]
> Would it help to document that somewhere?

It is, kinda: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials ;-)

**Comment by subotic** [comment]
If someone writes it's own services, then this would be a thing that they should know. If they use our library, then this will be already set correctly.

**Comment by gfoo** [comment]
Don't know if it is clear but I use `http://0.0.0.0:3333` as Knora base url to get cookie working from my Angular app, but as related in #1174 and #1130 my Sipi container for an unknown reason cannot access to `0.0.0.0`, set I use container name in docker compose file.

**Comment by subotic** [comment]
Yes, this is how it should work locally.

**Comment by gfoo** [comment]
yes, thanks.
